dbms/src/Interpreters/Join.cpp

```diff
@@ -1974,11 +1978,18 @@ std::optional<RestoreInfo> Join::getOneRestoreStream(size_t max_block_size_)
             // build new restore infos.
             auto spilled_partition_index = spilled_partition_indexes.front();
             RUNTIME_CHECK_MSG(hash_join_spill_context->isPartitionSpilled(spilled_partition_index), "should not restore unspilled partition.");
+
             if (restore_join_build_concurrency <= 0)
                 restore_join_build_concurrency = getRestoreJoinBuildConcurrency(partitions.size(), spilled_partition_indexes.size(), join_restore_concurrency, probe_concurrency);
-            /// for restore join we make sure that the build concurrency is at least 2, so it can be spill again
-            assert(restore_join_build_concurrency >= 2);
+            /// for restore join we make sure that the restore_join_build_concurrency is at least 2, so it can be spill again.
+            /// And restore_join_build_concurrency should not be greater than probe_concurrency, Otherwise some restore_stream will never be executed.
+            RUNTIME_CHECK_MSG(
```

Comment:
Yes, because each physical_plan will ensure that the concurrency of the output stream is max_threads, so the condition of runtime_check must be met here.
For pipeline model, https://github.com/pingcap/tiflash/pull/7833/files#diff-4cf628b56bfec1b29cc5eedafc0c2ff9e85900308d256dd7bb9331e6183e40d4R29-R35 can guarantee that the condition of runtimecheck can be met.